### PR TITLE
Big FireSquid is coming to the town 🔥🦑

### PR DESCRIPTION
**Pull request by vikiival**
tenor

**Thank you for your contribution** to the [KodaDot Indexer](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [x] Bugfix
- [x] Feature
- [x] Refactoring

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [ ] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

### Optional

- [x] I've run the indexer and it hasn't failed
- [x] I've changed schema and performed a migration
- [ ] I found edge cases

### Screenshot

Screenshot 2022-08-03 at 14 39 23

### CODE REVIEW

**Comment by vikiival** [comment]
cc @kngzhi

File: src/mappings/index.ts [hunk 1]
```diff
@@ -1,170 +1,156 @@
...
+    final.createdAt = timestamp
+    // final.events = [collectionEventFrom(RmrkEvent.MINT, remark, '')]
 
     // logger.watch(`[MINT] ${final.events[0]}`)
 
-    const metadata = await handleMetadata(final.metadata, final.name, store)
-    final.meta = metadata
+    if (final.metadata) {
+      const metadata = await handleMetadata(final.metadata, final.name, context.store)
+      final.meta = metadata
```
Comment (petersopko) [inline src/mappings/index.ts:120]:
```suggestion
      final.meta = await handleMetadata(final.metadata, final.name, context.store)
```

File: src/mappings/index.ts [hunk 2]
```diff
@@ -1,170 +1,156 @@
...
-    final.events = [collectionEventFrom(RmrkEvent.MINT, remark, '')]
+    final.createdAt = timestamp
+    // final.events = [collectionEventFrom(RmrkEvent.MINT, remark, '')]
 
     // logger.watch(`[MINT] ${final.events[0]}`)
 
-    const metadata = await handleMetadata(final.metadata, final.name, store)
-    final.meta = metadata
+    if (final.metadata) {
+      const metadata = await handleMetadata(final.metadata, final.name, context.store)
```
Comment (petersopko) [inline src/mappings/index.ts:119]:
```suggestion
      
```

File: src/mappings/index.ts [hunk 3]
```diff
@@ -1,170 +1,156 @@
...
     )
-    canOrElseError<CollectionEntity>(exists, collection, true)
-    isOwnerOrElseError(collection, remark.caller)
-    const final = create<NFTEntity>(NFTEntity, collection.id, {})
-    const id = getNftId(nft, remark.blockNumber)
+    plsBe(real, collection)
+    isOwnerOrElseError(collection, caller)
+    const id = getNftId(nft, blockNumber)
+    // const entity = await get<NFTEntity>(context.store, NFTEntity, id)
+    // plsNotBe<NFTEntity>(real, entity as NFTEntity)
```
Comment (petersopko) [inline src/mappings/index.ts:147]:
```suggestion

```

File: src/mappings/index.ts [hunk 4]
```diff
@@ -173,216 +159,197 @@ async function mintNFT(
...
-    final.updatedAt = remark.timestamp
-    // final.events = [eventFrom(RmrkEvent.MINTNFT, remark, '')]
+    final.createdAt = timestamp
+    final.updatedAt = timestamp
 
-    const metadata = await handleMetadata(final.metadata, final.name, store)
-    final.meta = metadata
+    if (final.metadata) {
+      const metadata = await handleMetadata(final.metadata, final.name, context.store)
+      final.meta = metadata
```
Comment (petersopko) [inline src/mappings/index.ts:167]:
```suggestion
      final.meta = await handleMetadata(final.metadata, final.name, context.store)
```

File: src/mappings/index.ts [hunk 5]
```diff
@@ -173,216 +159,197 @@ async function mintNFT(
...
+    plsBe<CollectionEntity>(real, collection)
+    isOwnerOrElseError(collection, caller)
+    collection.currentOwner = interaction.value
+    // collection.events?.push(
+    //   collectionEventFrom(
+    //     RmrkEvent.CHANGEISSUER,
+    //     remark,
+    //     ensure<string>(interaction.metadata)
+    //   )
+    // )
```
Comment (petersopko) [inline src/mappings/index.ts:309]:
```suggestion

```

File: src/mappings/utils/cache.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import logger, { logError } from './logger';
 import { Series, Spotlight, CacheStatus, Collector } from "../../model/generated";
```
Comment (petersopko) [inline src/mappings/utils/cache.ts:2]:
```suggestion
import { Series, Spotlight, CacheStatus, Collector } from "../../model";
```

File: src/mappings/utils/consolidator.ts [hunk 7]
```diff
@@ -1,22 +1,23 @@
-import { BatchArg, RmrkInteraction } from './types'
+import { BatchArg, ExtraCall, RmrkInteraction, Transfer } from './types'
 import { CollectionEntity, NFTEntity } from '../../model/generated'
```
Comment (petersopko) [inline src/mappings/utils/consolidator.ts:2]:
```suggestion
import { CollectionEntity, NFTEntity } from '../../model'
```

File: src/mappings/utils/consolidator.ts [hunk 8]
```diff
@@ -58,20 +67,19 @@ export function isPositiveOrElseError(entity: bigint | number, excludeZero?: boo
...
-//   } catch (e) {
-//     throw new ReferenceError(`[CONSOLIDATE Invalid account] ${caller}`)
-//   }
-// }
+// kodadot/rubick#6
+function paperCut({id}: NFTEntity, { remarkCount }: ExtraCall) {
+  if (remarkCount > 1) {
+    throw new ReferenceError(`[CONSOLIDATE] Entity: ${id} should have only one remark per batch, got: ${remarkCount}`)
+  }
+}
```
Comment (petersopko) [inline src/mappings/utils/consolidator.ts:85]: seems to be unused function?
```suggestion
```

File: src/mappings/utils/types.ts [hunk 9]
```diff
@@ -123,6 +113,21 @@ export type MetadataAttribute = {
   value: number | string
 }
 
+export type Transfer = {
+  to: any,
```
Comment (KngZhi) [inline src/mappings/utils/types.ts:117]: maybe avoid `any`, `string` would be better.

File: src/processor.ts [hunk 10]
```diff
@@ -1,18 +1,30 @@
...
+// good test values from start { from: 5756453, to: 5828351 } }
+// processor.addCallHandler('System.remark', { range: { from: 13445851, to: 13445851 } } , async ctx => {
+//     const kek = unwrap(ctx, getCreateToken)
+//     const extra = extractExtra(ctx)
+//     console.log('[NAME]', ctx.extrinsic.call.name)
+//     console.log('[EXTRA]', ctx.call.parent?.args?.calls)
+//     console.log('[EXTRA]', ctx.call.parent?.args?.calls?.at(0)?.value)
+//     console.log('[ARGS]', extra)
+//     console.log('[RMRK]', kek)
+// });
```
Comment (KngZhi) [inline src/processor.ts:28]: maybe remove that?
If we really need, could search on `git-log` of file.

File: src/mappings/utils/consolidator.ts (hunk 8 above)
Comment (vikiival) [reply src/mappings/utils/consolidator.ts:85]: Yup, it could break the state #6

File: src/mappings/utils/consolidator.ts (hunk 7 above)
Comment (vikiival) [reply src/mappings/utils/consolidator.ts:2]: This is autoimport

File: src/mappings/index.ts (hunk 3 above)
Comment (vikiival) [reply src/mappings/index.ts:147]: #5

**Pull request merged**